libfive/src/eval/eval_deriv_array.cpp

```diff
@@ -210,6 +210,24 @@ void DerivArrayEvaluator::operator()(Opcode::Opcode op, Clause::Id id,
         case Opcode::OP_ATAN:
             od = ad.rowwise() / (pow(av, 2) + 1);
             break;
+        case Opcode::OP_SINH:
+            od = ad.rowwise() * cosh(av);
+            break;
+        case Opcode::OP_COSH:
+            od = ad.rowwise() * sinh(av);
+            break;
+        case Opcode::OP_TANH:
+            od = ad.rowwise() * pow(1/cosh(av), 2);
+            break;
+        case Opcode::OP_ASINH:
+            od = ad.rowwise() / sqrt(pow(av, 2) + 1);
+            break;
+        case Opcode::OP_ACOSH:
+            od = ad.rowwise() / sqrt(pow(av, 2) - 1);
```

Comment:
According to [Wolfram Alpha](https://www.wolframalpha.com/input/?i=d%2Fdx+acosh%28f%28x%29%29), this should be `ad.rowwise() / (sqrt(av - 1) * sqrt(av + 1))`

(not sure why it's asymmetric with `asinh`, but [MathWorld](https://mathworld.wolfram.com/InverseHyperbolicCosine.html) says the same – maybe this is equivalent to what you write [if `f(x)` is positive](https://www.wolframalpha.com/input/?i=sqrt%28x+-+1%29+*+sqrt%28x+%2B+1%29)?)

Reply:
they are equivalent for x is positive. easy correction.